File: wallendar-app/app/src/test/java/com/td/wallendar/group/ui/GroupPresenterTest.java
package com.td.wallendar.group.ui;

import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertFalse;
import static org.junit.Assert.assertTrue;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.when;
import static io.reactivex.Single.error;
import static io.reactivex.Single.just;

import com.td.wallendar.TestSchedulerProvider;
import com.td.wallendar.models.ApplicationUser;
import com.td.wallendar.models.Group;
import com.td.wallendar.repositories.interfaces.GroupsRepository;
import com.td.wallendar.utils.scheduler.SchedulerProvider;

import org.junit.Before;
import org.junit.Test;

import java.util.Collections;
import java.util.HashSet;

public class GroupPresenterTest {

    private GroupView groupView;
    private GroupsRepository groupsRepository;
    private SchedulerProvider schedulerProvider = new TestSchedulerProvider();

    private GroupPresenter presenter;

    private final Group fakeGroup = new Group(1L, "group title", new ApplicationUser(), new HashSet<>(), new HashSet<>(), new HashSet<>(), new HashSet<>());

    @Before
    public void setup() {
        groupView = mock(GroupView.class);
        groupsRepository = mock(GroupsRepository.class);

        presenter = new GroupPresenter(groupView, groupsRepository, schedulerProvider);
    }

    @Test
    public void givenAnUserIdAndTheViewAttached_ThenTheViewShouldShowTheGroupInformation_Ok() {
        when(groupsRepository.getGroup(1L)).thenReturn(just(fakeGroup));

        presenter.onViewAttached(1L);

        verify(groupView).bindGroup(fakeGroup);
        verify(groupView).bindGroupHistory(Collections.emptyList());
    }

    @Test
    public void givenAnUserIdAndTheViewAttached_ThenTheViewShouldShowTheGroupInformation_WithError() {
        when(groupsRepository.getGroup(1L)).thenReturn(error(new Exception("Error while fetching group")));

        presenter.onViewAttached(1L);

        verify(groupView).onGroupLoadError();
    }

    @Test
    public void whenViewIsDetached_ThenDisposableShouldBeCleared() {
        presenter.onViewDetached();

        assertEquals(0, presenter.getDisposable().size());
    }

    @Test
    public void whenViewIsDestroyed_ThenDisposableShouldBeDisposed() {
        presenter.onViewDetached();

        assertFalse(presenter.getDisposable().isDisposed());
    }
}

File: wallendar-app/app/src/test/java/com/td/wallendar/group/ui/GroupPresenterTest.kt
package com.td.wallendar.group.ui

import com.td.wallendar.TestSchedulerProvider
import com.td.wallendar.models.Group
import io.reactivex.Single
import org.junit.Assert
import org.junit.Before
import org.junit.Test
import java.lang.Exception
import java.util.HashSet

class GroupPresenterTest {
    private var groupView: GroupView? = null
    private var groupsRepository: GroupsRepository? = null
    private val schedulerProvider: SchedulerProvider? = TestSchedulerProvider()
    private var presenter: GroupPresenter? = null
    private val fakeGroup: Group? = Group(1L, "group title", ApplicationUser(), HashSet(), HashSet(), HashSet(), HashSet())
    @Before
    fun setup() {
        groupView = Mockito.mock<GroupView?>(GroupView::class.java)
        groupsRepository = Mockito.mock<GroupsRepository?>(GroupsRepository::class.java)
        presenter = GroupPresenter(groupView, groupsRepository, schedulerProvider)
    }

    @Test
    fun givenAnUserIdAndTheViewAttached_ThenTheViewShouldShowTheGroupInformation_Ok() {
        Mockito.`when`<Single<Group?>?>(groupsRepository.getGroup(1L)).thenReturn(Single.just(fakeGroup))
        presenter.onViewAttached(1L)
        Mockito.verify<GroupView?>(groupView).bindGroup(fakeGroup)
        Mockito.verify<GroupView?>(groupView).bindGroupHistory(emptyList())
    }

    @Test
    fun givenAnUserIdAndTheViewAttached_ThenTheViewShouldShowTheGroupInformation_WithError() {
        Mockito.`when`<Single<Group?>?>(groupsRepository.getGroup(1L)).thenReturn(Single.error<Group?>(Exception("Error while fetching group")))
        presenter.onViewAttached(1L)
        Mockito.verify<GroupView?>(groupView).onGroupLoadError()
    }

    @Test
    fun whenViewIsDetached_ThenDisposableShouldBeCleared() {
        presenter.onViewDetached()
        Assert.assertEquals(0, presenter.getDisposable().size().toLong())
    }

    @Test
    fun whenViewIsDestroyed_ThenDisposableShouldBeDisposed() {
        presenter.onViewDetached()
        Assert.assertFalse(presenter.getDisposable().isDisposed)
    }
}